### PR TITLE
feat: Sync on join

### PR DESCRIPTION
**Pull request by LukeHankey**
This PR adds the sync event and refactors some methods in ConnectionManager, handing them over to the EventHandler

### CODE REVIEW

File: server/connection_manager.py [hunk 1]
```diff
@@ -100,19 +85,19 @@ def update_code_cache(self, room_code: str, replace_data: ReplaceData) -> None:
...
             data: The data to be sent to the clients.
             room_code: The room to which the data will be sent.
-            sender (optional): The client who sent the message.
+            sender (optional): The client who sent the request.
             buggy (optional): To send back modified code.
         """
         if buggy:
             data = self._modify_code(room_code)
-            self.update_code_cache(room_code, data)
+            self.update_code_cache(room_code, data.data)
```
Comment (kappq) [inline server/connection_manager.py:100]: `data.data` looks a bit weird. :laughing:

File: server/events.py [hunk 2]
```diff
@@ -204,6 +220,25 @@ def __call__(self, request: EventRequest, room_code: str) -> tuple[bool, Client
...
+                await self.client.send(EventResponse(type=EventType.SYNC, data=data, status_code=StatusCode.SUCCESS))
+                connect_data = cast(
+                    ConnectData,
+                    {
+                        "connection_type": "join",
+                        "difficulty": None,
+                        "room_code": room_code,
+                        "username": self.client.username,
+                    },
+                )
```
Comment (kappq) [inline server/events.py:239]: The `cast` method shouldn't be required, the data could be represented like so:
```py
ConnectData(connection_type="join", room_code=room_code, username=self.client.username)
```
which, in my opinion, looks more clear and avoids typing errors.

File: server/connection_manager.py (hunk 1 above)
Comment (LukeHankey) [reply server/connection_manager.py:100]: It is what it is xD

File: server/events.py (hunk 2 above)
Comment (LukeHankey) [reply server/events.py:239]: Type checkers don't like this which is why I casted the EventData to each of their actual types. The casts also provided me with intellisense. The cast is only used to appease the type checkers

**Review by kappq — APPROVED** [review]

**Pull request merged**